RFD 83: Host cert support in Machine ID

timothyb89 opened the pull request:
This RFD describes changes needed to support issuing OpenSSH host certificates via Machine ID. In particular, we want to support restricting host certificate issuance via RBAC rules, and provide some UX for users to request and update these certificates.

timothyb89 committed:
RFD 83: Host cert support in Machine ID

This RFD describes changes needed to support issuing OpenSSH host
certificates via Machine ID.

strideynet commented on rfd/0083-machine-id-host-certs.md:
```diff
@@ -0,0 +1,150 @@
+renewals are today.
+
+Open questions: bots still renew all certs at startup. Do we maintain that
+behavior for these? If not, how do we know when to renew certs next?
```
I suppose one option is to read the host certs and their issuance/expiry time out of them if the exist, and use that to determine a threshold after which we should renew them.

timothyb89 replied:
Yeah, I think that's what would need to happen if we go this route.

I think my next step will be testing OpenSSH's behavior to see if we can just refresh its certs every 20 minutes without breaking things.

timothyb89 replied:
closing the loop on this, there doesn't seem to be any real downside to frequent rotations:
* sshd reloads certs at login time, no explicit reload needed
  * even if we did need to reload sshd, ongoing sessions are still not impacted and keep running smoothly
* neither expired nor reloaded certs impact ongoing sessions, it's only ever checked at first connect
* clients starting new sessions automatically accept the new certs so long as they use our generated `known_hosts` which includes the cluster CA

All told, there's no real downsides to frequent cert rotation, though I did document a potential UX issue that could arise in the RFD. It's hopefully minor enough that we won't need to address it.

timothyb89 committed:
Small typo fix

timothyb89 commented on rfd/0083-machine-id-host-certs.md:
```diff
@@ -0,0 +1,150 @@
+  - directory: /opt/machine-id
+    configs:
+      - host_cert:
+          principals: [foo.example.com]
```
That's a good question, I think we could do that (or provide it via some special word). It does run into the same problems though - malicious hosts could request certs to impersonate another server. I bet as with https://github.com/gravitational/teleport/issues/13483 we could also grant a default allow exception for the client's own IP?

Otherwise, by default, the `host_cert` rule doesn't have a `where` clause so bots with the rule can issue whichever certs they'd like. If users want to lock this down they'll need some predictable pattern to the principals, whether they're IPs or DNS names. We can use `starts_with()` or `matches()` rules (or a combination) to match against all possibilities.

strideynet replied:
Yeah, there's some work we need to do on #15521 that will let the auth service see the client's real address, then we could at least have a special keyword or similar to allow this.

timothyb89 committed:
Update RFD with additional sshd UX findings

timothyb89 committed:
Small RFD tweaks based on prototype findings

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/machine-id-host-cert-rfd

timothyb89 committed:
Add more detail on short lived certs and where predicate fields

timothyb89 committed:
Clarify new predicate function intent

zmb3 commented on rfd/0083-machine-id-host-certs.md:
```diff
@@ -0,0 +1,198 @@
+2. Ensuring these functions support lists of strings (or building separate
+   functions specifically for matching against `[]string`).
+
+   For example, a hypothetical `ends_with(principals, ".example.com")` should 
```
Can you think of any valid use cases for an OR instead of an AND?

I ask because we're getting lots of requests to support more OR conditions in our RBAC engine lately.

We could instead have `some_end_width(principals, suffix)` and `all_end_with(principals, suffix)` if we need greater flexibility.

timothyb89 replied:
That's a good thought, and that naming scheme opens the door for a few extra functions like `all_equal()`. That said, I think `some_end_with()` would probably be pretty dangerous in most RBAC situations since it effectively allows anything through.

We don't have a great syntax for "allow any subset of these options" right now. It is possible (albeit ugly) to use something like:

```
all_end_with(host_cert.principals, ".foo.example.com") || all_end_with(host_cert.principals, ".bar.example.com")
```

... but this wouldn't allow requesting _both_ `node.foo.example.com` and `node.bar.example.com`. Ultimately we'd need to specify a list / varargs and I don't believe the `predicate` library has any support for that syntactically. I'm not sure there's a great way forward here aside from a regex matcher (which has its own problems).

zmb3 commented on rfd/0083-machine-id-host-certs.md:
```diff
@@ -0,0 +1,198 @@
+expiration date][date]. While we could preserve this behavior, any process that
+continually produces certificates with an infinite TTL seems problematic.
+
+Our preference is to generate similarly short-lived certificates as we do for
```
Strongly agree.

zmb3 commented on rfd/0083-machine-id-host-certs.md:
```diff
@@ -0,0 +1,198 @@
+does not care about file permissions unless they're owned by `root`.
+
+A downside to this approach is that permissions errors (for example, if the
+`where` predicate doesn't match) will be reported as errors but will not crash
```
Should we emit audit events for these errors?

timothyb89 replied:
It is logged today (in debug mode only) but not as an audit event. I think this is standard Teleport RBAC behavior? I don't see any events emitted for any failed RBAC checks, surprisingly.

strideynet replied:
It feels like we certainly should be auditing failed RBAC checks, especially for generating certificates ! If you feel its out of scope for this, can we at least raise an issue to add an audit event for this ?

timothyb89 replied:
We do have rule actions which can produce logs (though not audit logs) on the off chance someone wants that for a particular rule. That might be an okay way forward, maybe.

strideynet replied:
Yes, perhaps not as useful for customers who are only ingesting the audit log into their SIEM, but probably the best we can do for now as part of this work.

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/machine-id-host-cert-rfd

timothyb89 committed:
Final batch of small tweaks based on feedback

timothyb89 commented:
Made a few final tweaks before merging: added a "future considerations" section, added an "all_equal()" matcher, and renamed "ends_with()" to "all_end_with()".
